Prolongs dfg graphs to the root of a language feature

konradweiss opened the pull request:
Currently we have the root node as branching node to keep the semantic information of the language-feature that branches while traversing the EOG. To keep EOG and DFG coherent and later build a propper PDG we prolong the dfg paths to the branching node.

I added the DFG edges where I added the EOG-Edges to the root node: in the EOG-pass- A strategy pattern would be nice to make the root node configurable for EOG and DFG but so far I could not find an easy way to integrate one for the EOG pass.

Root node of language feature if is a sink for the EOG of the condition and then branches out. For certain applications this rootnode allows traversals to use the root node like a jumpi 
![eog_root_node](https://user-images.githubusercontent.com/12885000/105500200-dc0a3780-5cc2-11eb-90ec-b8c70b8f65a9.png)

DFG edge from condition to the language-feature root node that acts as branching condition.
![dfg_root_node](https://user-images.githubusercontent.com/12885000/105500198-db71a100-5cc2-11eb-876c-26af83ca867b.png)

konradweiss commented:
I also tested this with codyze and the tests run through so it should be fine

oxisto commented on src/main/java/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.java:
```diff
@@ -781,7 +782,10 @@ private void handleDoStatement(@NonNull Node node) {
 
     createEOG(doStatement.getCondition());
+
+    doStatement.addPrevDFG(doStatement.getCondition());
```
I am semantically confused about an evaluation order pass, that adds data flow edges. Does that make sense?

oxisto commented:
> Currently we have the root node as branching node to keep the semantic information of the language-feature that branches while traversing the EOG. To keep EOG and DFG coherent and later build a propper PDG we prolong the dfg paths to the branching node.
> 
> 
> 
> I added the DFG edges where I added the EOG-Edges to the root node: in the EOG-pass- A strategy pattern would be nice to make the root node configurable for EOG and DFG but so far I could not find an easy way to integrate one for the EOG pass.
> 
> 
> 
> Root node of language feature if is a sink for the EOG of the condition and then branches out. For certain applications this rootnode allows traversals to use the root node like a jumpi 
> 
> ![eog_root_node](https://user-images.githubusercontent.com/12885000/105500200-dc0a3780-5cc2-11eb-90ec-b8c70b8f65a9.png)
> 
> 
> 
> DFG edge from condition to the language-feature root node that acts as branching condition.
> 
> ![dfg_root_node](https://user-images.githubusercontent.com/12885000/105500198-db71a100-5cc2-11eb-876c-26af83ca867b.png)
> 
> 
> 
> 

Could you provide the code associated with the example graph? I am still not quite sure if I understand it correctly.

konradweiss commented:
> > Currently we have the root node as branching node to keep the semantic information of the language-feature that branches while traversing the EOG. To keep EOG and DFG coherent and later build a propper PDG we prolong the dfg paths to the branching node.
> > I added the DFG edges where I added the EOG-Edges to the root node: in the EOG-pass- A strategy pattern would be nice to make the root node configurable for EOG and DFG but so far I could not find an easy way to integrate one for the EOG pass.
> > Root node of language feature if is a sink for the EOG of the condition and then branches out. For certain applications this rootnode allows traversals to use the root node like a jumpi
> > ![eog_root_node](https://user-images.githubusercontent.com/12885000/105500200-dc0a3780-5cc2-11eb-90ec-b8c70b8f65a9.png)
> > DFG edge from condition to the language-feature root node that acts as branching condition.
> > ![dfg_root_node](https://user-images.githubusercontent.com/12885000/105500198-db71a100-5cc2-11eb-876c-26af83ca867b.png)
> 
> Could you provide the code associated with the example graph? I am still not quite sure if I understand it correctly.

`
if(num > maxSoFar){
...
}
`
